### ISSUE TITLE
Online code

### ISSUE THREAD

**Issue by zhujiagang** [gurkirt/realtime-action-detection#2]
I've run and read your code, an excellent work. Have you realized your online real-time testing code, that is, given a video, frames are fed into the model by order and corresponding detection results are displayed including bbx and predicted label? Vkalogeiton [#1](https://github.com/vkalogeiton/caffe/tree/act-detector) has released his code totally by python but without online version, I plan to modify his code to realize online deploying.

**Comment by gurkirt** [comment]
Linking part is completely online https://github.com/gurkirt/realtime-action-detection/blob/master/online-tubes/actionpath/incremental_linking.m; you only need to put together temporal labeling with from https://github.com/gurkirt/realtime-action-detection/blob/master/online-tubes/gentube/dpEM_max.m; 

What you are asking a demo style code; I am sorry to say that I didn't had time to make a deployable code. you can take inspiration from  Vkalogeiton code. She has it python.If you do get the time to convert  online linking into deployable code then please do share. I will put a link or I can accept a PR

**Comment by zhujiagang** [comment]
Thank you very much, I will try. Wait for my good news.

**Comment by shijieS** [comment]
Hi, jianggang. Have finished the online real-time testing code?

**Comment by zhujiagang** [comment]
@shijieS I am currently working on it. Firstly, I decide to modify its matlab code in `I01onlineTubes` to realize online detection by offline features. Because I found that the code needs some adaption. Then based on understanding, it is possible to convert it into python because its features are extracted by pytorch.

**Comment by gurkirt** [comment]
I would recommend converting `incremental_liniking.m` to python as the first thing. NMS and other masking stuff are already there in `test-ucf24.py`. It would easier to plug everything together and make demo ready code. Also, `dpEM.m` already has python version in one of other repos https://github.com/gurkirt/actNet-inAct/blob/master/processing/detection-fusion-working.py#L146. I will push that in this repo soon. Only change require there that `dpEM.m` solve for max cost and python version solve for min cost.

**Comment by zhujiagang** [comment]
I find some possible bug in [](https://github.com/gurkirt/realtime-action-detection/blob/b2f48372d3b893191c88acd54dada7988c1e5f7c/online-tubes/actionpath/incremental_linking.m#L54).
It perhaps should be `live_paths(lp).lastfound = live_paths(lp).lastfound +1;`

**Comment by gurkirt** [comment]
I think you are right. I will run the experiments again and report back.
Thanks

**Comment by zhujiagang** [comment]
I have made a demo that could display bbx and prediction score for streaming frames. But it (by matlab) takes the offline bbx and prediction score as the input. Before I send a pull request and move forward to write a total python version, I want to ask a question about the online viterbi: when and how does the pairwise potential `alpha = 3`  work in a tube?  I found the online vertibi similar to a summing operation, though the `alpha = 3` sometimes work, it couldn't change the prediction label of the whole tube.

**Comment by gurkirt** [comment]
Great! Thanks for your effort!
Alpha is being used in Viterbi in `online-tubes/gentube/dpEM_max.m`. Once you have paths form `incremental_liniking.m` then you pass the allScores variable of each path through `dpEM`.  Interface is given in `PARactionpathsmoother.m`. Python version which minimizes rather than maximization is there in one of my other repo https://github.com/gurkirt/actNet-inAct/blob/master/processing/detection-fusion-working.py#L146.

GOOD LUCK!

**Comment by zhujiagang** [comment]
As a simple example, `p1`, `p2` and `p3` denote the prediction sequence of three actions, and `D1`, `D2` and `D3` denote the corresponding row of `D` matrix in [https://github.com/gurkirt/realtime-action-detection/blob/b2f48372d3b893191c88acd54dada7988c1e5f7c/online-tubes/gentube/dpEM_max.m#L47](url). Suppose `p1, p2, p3` equal to `0.6, 0.3, 0.1` respectively for all time steps. Then according to your code, calculating process is just as follows:
```
      t    t+1 ... t+10   t+11                                                 t+12
D1 0.6  1.2  ... 6.6      7.2                                                     7.8
p1  0.6  0.6  ... 0.6     0.6                                                      0.6

D2 0.3 0.6  ...   3.3     0.3 + max(6.6-alpha (3), 3.3) = 3.9      0.3 + max(7.2-alpha(3), 3.9) = 4.5
p2 0.3 0.3   ...  0.3     0.3                                                       0.3

D3  0.1 0.2 ...
p3  0.1 0.1 ...
```

when time step is `t+11`, `alpha` starts to work and it could prevent the gap between `D1` and `D2` from being larger than a threshold, for example, `t+11, 7.2 - 3.9 = 3.3, t+12 7.8 - 4.5 = 3.3.`
`D1` will always be larger than `D2`, unless `p2` is larger than `p1` in coming frames. But I think in a tube, the prediction scores are always stable and the score of one action is always larger than other actions'.  I couldn't imagine a case where `alpha` could truly work in label smoothing.

Second example
```
     t      t+1 ...
D1  0.5  0.9 ...
p1  0.5  0.4 ...

D2  0.4 0.85 ...
p2  0.4 0.45  ...
```

In this case, both `p1` and `p2` are close to `0.5`, in this case `alpha` will not work, because the difference between `D1` and `D2` is smaller than `alpha`. So the online viterbi degenerates to a summing operation.

So I find that the online vertibi is similar to a summing operation, becuase `alpha` seldom works. Could you give more examples to illustrate this algorithm?

**Comment by gurkirt** [comment]
I agree, I lot of cases it won't work but it works in the cases where clear evidence in scores sequence that sequences should be trimmed. The logic here is that we don't want to switch the label very frequently hence the higher penalty (0.5) if the sum of difference goes more than the penalty(alpha) over few frames then you should be changing the label. It is a conservative strategy, which keeps the label constant for few frames even the evidence points otherwise but not very strongly, plus you don't want the change the label immediately if score falls below 0.5, that would result in greedy thresholding, which will result in the sequence with a lot of abrupt holes. 

As for the above example. There is no reason to switch from p1 to p2 as p1 always higher than p2 and the p1 track doesn't have to pay any cost to stay on the track of p1. At some point of p2 becomes very high, say 0.7, then track form D1 would switch to p1 and p2 track. In your example, after some point D1, D2 and D3 will converge to D1 track, when you backtrace, you will find them in 2nd and 3rd optimal paths.

I hope it helps

**Comment by zhujiagang** [comment]
Yes, I got it. I've send a pr.

**Comment by gurkirt** [comment]
I saw the PR. I see there aren't any changes existing files, you create a new directory and pushed those?

**Comment by zhujiagang** [comment]
Yes, I add a new subdirectory `matlab-online-display` to contain all the needed files, because some of these files are changed and are no longer compatible with your original tube generation files.

**Comment by gurkirt** [comment]
Okay, I will accept it as it is.

**Comment by gurkirt** [comment]
Thanks  a lot :)

**Comment by gurkirt** [comment]
I will make amends to README.md to mention your contribution.

**Comment by zhujiagang** [comment]
Thank you! This is my first PR on git! Could you run the demo without errors?

**Comment by Feynman27** [comment]
@zhujiagang Thanks for the PR! Can you summarize how the source code in ```matlab-online-dispay``` is different than those in ```online-tubes```? It's not exactly obvious to me from the discussion above.

**Comment by zhujiagang** [comment]
1 For demo, change the video list to `display01.txt` by including only two videos, which could be seen from `matlab-online-display/utils/initDatasetOptsFused.m, initDatasetOpts.m`.
2 move the label smoothing of `gettubes` to `actionPaths` to realize online smoothing, delete `gettubes` in `I01onlineTubes.m`, `I02genFusedTubes.m`.
3 do nms (`dofilter`) on one frame (`actionPaths.m`), then linking, online viterbi (`incremental_linking.m`), then loop,
instead of do nms on all frames.
4 add `dp_count` to struct `my_dead_paths`, `D, phi` to struct `live_paths`, to remember these variables during loop by struct.
5 add display code, when all classes are visited, set candidate display threshold `[0.01, 0.5, 0.9]`.

**Comment by gurkirt** [comment]
There are FEW mistakes in current readme and training strategy for RGB model. I will publish all this together. Hopefully, I will finish all the experiments by tomorrow.

**Comment by Feynman27** [comment]
I pushed a python version of the incremental_linking to my fork at: https://github.com/Feynman27/realtime-action-detection

Tested on a single video, and `actionpaths` match. Also, noticed that if I use the `nms` function from py-faster-rcnn, I sometime get a subset of filtered boxes. Using your pytorch implementation is the only way to reproduce the `actionpaths` from the matlab code.

**Comment by gurkirt** [comment]
Thanks a lot for your effort. I will try to merge it in next release. 

I am not sure if you are implying that nms that I am using is different than that in `py-faster-RCNN`?

Can point me where you made the changes?

**Comment by Feynman27** [comment]
I'm going to convert the [`getTubes`](https://github.com/Feynman27/realtime-action-detection/blob/264eed0b218629e1fdeb4957bff0c25687ebabcf/online-tubes/I01onlineTubes.m#L65-L66) function before merging. The changes are in commit [330f6d4dde0116d27bc2f251bdbd495bf9878fe4](https://github.com/Feynman27/realtime-action-detection/commit/330f6d4dde0116d27bc2f251bdbd495bf9878fe4) and [264eed0b218629e1fdeb4957bff0c25687ebabcf](https://github.com/gurkirt/realtime-action-detection/commit/264eed0b218629e1fdeb4957bff0c25687ebabcf). 

Regarding the nms, I'm not sure why the nms is py-faster-rcnn is removing more boxes than the current torch implementation. I'll try to find time to look into it more and create a new issue if need be.
